### PR TITLE
Add support for the updated import attributes proposal

### PR DESCRIPTION
**Pull request by nicolo-ribaudo**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | 
| Patch: Bug Fix?          |
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR Link    | https://github.com/babel/website/pull/2780
| Any Dependency Changes?  |
| License                  | MIT

https://github.com/tc39/proposal-import-attributes

The changes compared to the import assertions proposals are listed at https://github.com/tc39/proposal-import-attributes/pull/131

To make the migration easier, since import assertions are already widely used in tools:
- the old `importAssertions` parser plugin accepts the new `with` syntax by default
- the new `importAttributes` parser plugin has a `deprecatedAssertSyntax` option to accept the old syntax

In Babel 8, we will remove both the `importAssertions` plugin and the super old `moduleAttributes` plugin (which used the `with` keyword, but without braces). I will add the proper `process.env.BABEL_8_BREAKING` checks in a future cleanup PR.

<details>
<summary>Description generated by Copilot 🤖:</summary>

### Summary
🆕🛠️🚧

This pull request adds support for the `importAttributes` plugin in the parser and the generator, and updates the code and tests to handle the new and legacy syntaxes for module attributes and assertions. It also renames the `importAssertions` plugin to `importAttributes` and adds a deprecation warning for the old name. It fixes some bugs and improves the error messages and the test output format for the related features.

> _Sing, O Muse, of the mighty pull request_
> _That brought new features to the Babel code,_
> _The `importAttributes` plugin, best_
> _Of all the tools that parse the module node._

### Walkthrough
*  Rename `importAssertions` plugin to `importAttributes` and update syntax to match proposal ([link](https://github.com/babel/babel/pull/15536/files?diff=unified&w=0#diff-0bc9bdcf377c7d34d3e83db3cf0a7df2b7b910c1f4fc20f1a99efcbd2c5b7643L87-R90), [link](https://github.com/babel/babel/pull/15536/files?diff=unified&w=0#diff-0bc9bdcf377c7d34d3e83db3cf0a7df2b7b910c1f4fc20f1a99efcbd2c5b7643R210-R216), [link](https://github.com/babel/babel/pull/15536/files?diff=unified&w=0#diff-3b38bf77133a2f29b9f9c1b31ed7b614745edff4daf532dec820b773414d321eR20-R27), [link](https://github.com/babel/babel/pull/15536/files?diff=unified&w=0#diff-3b38bf77133a2f29b9f9c1b31ed7b614745edff4daf532dec820b773414d321eR99-R111), [link](https://github.com/babel/babel/pull/15536/files?diff=unified&w=0#diff-3b38bf77133a2f29b9f9c1b31ed7b614745edff4daf532dec820b773414d321eR172), [link](https://github.com/babel/babel/pull/15536/files?diff=unified&w=0#diff-c970b1ea6dab92947182a2db40691bfd5f8a7a281c1aeb88a8bacf945b4a9bf4L177-R185))
*  Add support for both `with` and `assert` keywords and both legacy and new module attribute syntaxes in `importAttributes` plugin ([link](https://github.com/babel/babel/pull/15536/files?diff=unified&w=0#diff-dde20a4c39ac9fc963a227ec21025e1af1368f92d8c0d1db85274b199adb1cf4L3009-R2996), [link](https://github.com/babel/babel/pull/15536/files?diff=unified&w=0#diff-dde20a4c39ac9fc963a227ec21025e1af1368f92d8c0d1db85274b199adb1cf4R3035-R3036), [link](https://github.com/babel/babel/pull/15536/files?diff=unified&w=0#diff-dde20a4c39ac9fc963a227ec21025e1af1368f92d8c0d1db85274b199adb1cf4L3055-R3044), [link](https://github.com/babel/babel/pull/15536/files?diff=unified&w=0#diff-dde20a4c39ac9fc963a227ec21025e1af1368f92d8c0d1db85274b199adb1cf4L3096-R3146), [link](https://github.com/babel/babel/pull/15536/files?diff=unified&w=0#diff-c970b1ea6dab92947182a2db40691bfd5f8a7a281c1aeb88a8bacf945b4a9bf4R202-R209))
*  Update `babel-generator` to handle both `attributes` and `assertions` properties on import and export nodes and use `_printAttributes` function ([link](https://github.com/babel/babel/pull/15536/files?diff=unified&w=0#diff-f428923c80a568eda61faa6081f14b7901fe84c4ae61f9289500dcc0b19827c1L69-R87), [link](https://github.com/babel/babel/pull/15536/files?diff=unified&w=0#diff-f428923c80a568eda61faa6081f14b7901fe84c4ae61f9289500dcc0b19827c1L96-R109), [link](https://github.com/babel/babel/pull/15536/files?diff=unified&w=0#diff-f428923c80a568eda61faa6081f14b7901fe84c4ae61f9289500dcc0b19827c1L176-R187), [link](https://github.com/babel/babel/pull/15536/files?diff=unified&w=0#diff-f428923c80a568eda61faa6081f14b7901fe84c4ae61f9289500dcc0b19827c1L254-R268))
*  Add error message for deprecated `assert` keyword in import attributes ([link](https://github.com/babel/babel/pull/15536/files?diff=unified&w=0#diff-9f39041ab8dd3240576e7f99020294ef8611f049a260c0d98580fd343322e373R114-R115), [link](https://github.com/babel/babel/pull/15536/files?diff=unified&w=0#diff-dde20a4c39ac9fc963a227ec21025e1af1368f92d8c0d1db85274b199adb1cf4L3009-R2996))
*  Update `ExportNamedDeclaration` and `ImportDeclaration` functions to use `maybeParseImportAttributes` function and remove unused functions ([link](https://github.com/babel/babel/pull/15536/files?diff=unified&w=0#diff-dde20a4c39ac9fc963a227ec21025e1af1368f92d8c0d1db85274b199adb1cf4L2598-R2599), [link](https://github.com/babel/babel/pull/15536/files?diff=unified&w=0#diff-dde20a4c39ac9fc963a227ec21025e1af1368f92d8c0d1db85274b199adb1cf4L2947-R2944))
*  Add test cases for `importAttributes` plugin with `with` keyword and `deprecatedAssertSyntax` option ([link](https://github.com/babel/babel/pull/15536/files?diff=unified&w=0#diff-2c3f0c9cd02f9138cb7f83a376c79b978ab078d9c28fc29ed7beb32e4520e87aL1-R1), [link](https://github.com/babel/babel/pull/15536/files?diff=unified&w=0#diff-fda339a848b510e034041c0b68785bba0d14ec809e8dad703d09d0d537e87e7fR1-R3), [link](https://github.com/babel/babel/pull/15536/files?diff=unified&w=0#diff-2a6ddb154d2561cc73e95c71f22a729f0f64a59351e5503d40429134b27c3702L1-R1))
*  Add test cases for `importAssertions` plugin with `assert` keyword ([link](https://github.com/babel/babel/pull/15536/files?diff=unified&w=0#diff-f17fc73f0f140622bb04f1bd0701fe0ba39a99683b327481c317b0cdbe7a6dc5L1-R1), [link](https://github.com/babel/babel/pull/15536/files?diff=unified&w=0#diff-c77b2f83427afecad2f95295a774ecb7a96624617a1746c60e9efc074455afadR1-R3), [link](https://github.com/babel/babel/pull/15536/files?diff=unified&w=0#diff-2a3427fbef355237ad0c2400c14cf2ad05df3a717031779494090aaca2bb8b21L1-R1))
*  Add test case for experimental syntax without any plugin enabled ([link](https://github.com/babel/babel/pull/15536/files?diff=unified&w=0#diff-795a41d2ed7d1bb088ce9ebd6f4d3cdb2e5a58e3df754020f8c76535150f2319R1), [link](https://github.com/babel/babel/pull/15536/files?diff=unified&w=0#diff-a5c094ee250f0dc2f72d0f4db87f9e8d45492719968b0f2eb0c5aac5a7076545R1-R4))
*  Add test case for `importAssertions` plugin with `with` keyword in dynamic import ([link](https://github.com/babel/babel/pull/15536/files?diff=unified&w=0#diff-230c6ab4b26430d2c9c0c4acf7dfc145ba612a55fe888924f1c482867e64a78eR1))
*  Update test output for `estree` plugin with import assertions and attributes ([link](https://github.com/babel/babel/pull/15536/files?diff=unified&w=0#diff-ba9c787f797bff3b77891fa9ce4fb61505a5314d08e9d025af3b4ad76580c1ffL2-R7), [link](https://github.com/babel/babel/pull/15536/files?diff=unified&w=0#diff-073bf5f314caa5170bcd6d3cabdcd85838ff8ae7ced4014ee93f43d66370dee8R31-R33), [link](https://github.com/babel/babel/pull/15536/files?diff=unified&w=0#diff-073bf5f314caa5170bcd6d3cabdcd85838ff8ae7ced4014ee93f43d66370dee8L51-R54))
*  Update test options for experimental syntax without any plugin enabled ([link](https://github.com/babel/babel/pull/15536/files?diff=unified&w=0#diff-3611ec52767d234d878e2bc0b82e809d2e3bb5ec3f723cd9722dee900eb68a73L2-R4))
*  Delete test files for deprecated `moduleAttributes` plugin ([link](https://github.com/babel/babel/pull/15536/files?diff=unified&w=0#diff-6fcafcbb0f6b13e136db8e5a024c5aa72f031786a93cb6af978dce347d039fd7), [link](https://github.com/babel/babel/pull/15536/files?diff=unified&w=0#diff-b761fa00b012f5e32581cd305fd4e3979fee7c3adb6f7e6c129c41f22aac04ea))

</details>

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/15536"></a>

### CODE REVIEW

File: packages/babel-parser/test/fixtures/experimental/import-assertions-with-keyword/linebreak-before-with/input.js [hunk 1]
```diff
@@ -0,0 +1,2 @@
+import "foo"
+with { type: "json" }
```
Comment (JLHwung) [inline packages/babel-parser/test/fixtures/experimental/import-assertions-with-keyword/linebreak-before-with/input.js:2]: The current spec disallows such case: https://tc39.es/proposal-import-attributes/#prod-ImportDeclaration

File: packages/babel-parser/src/parser/statement.ts [hunk 2]
```diff
@@ -3087,27 +3069,81 @@ export default abstract class StatementParser extends ExpressionParser {
...
+    }
+  }
+
+  maybeParseImportAttributes(
+    node: Undone<N.ImportDeclaration | N.ExportNamedDeclaration>,
+  ) {
+    let attributes: N.ImportAttribute[];
+    let useWith = false;
+
+    if (this.match(tt._with)) {
```
Comment (JLHwung) [inline packages/babel-parser/src/parser/statement.ts:3090]:
```suggestion
    // https://tc39.es/proposal-import-attributes/#prod-WithClause
    if (this.match(tt._with)) {
```
Nit. Attaching the spec link here for reference.

**Review by JLHwung — COMMENTED** [review]
Can we introduce a new generator option 
```ts
importAttributesVersion: "2023-03" | "2020-09" | "2020-06"
```
The default is `"2023-03"`. So we can handle the AST collision between import attributes and the legacy module attributes proposal. Then syntax plugins can configure it accordingly.

The pros are:
- Sigil configs are decoupled from the AST, so we don't have to plug in more private-like `node.extra.*` properties
- Users don't need to write a dedicate codemod to migrate syntaxes (by manipulating `node.extra.*` properties)
- The output style is controlled in generator option instead of a parser plugin option, e.g. `importAttributesVersion: "2020-09"` can be used for Node.js 17.1.0 compatibility, though we will eventually materialize that option after import attributes gets spec'd.

The cons:
For users using the `babel-plugin-syntax-module-attributes` in the small time frame 2020-06 to 2020-09 (prior to renaming module-attributes to import-assertions), upgrading to Babel 7.22 will be breaking, unless they configured `importAttributesVersion: "2020-06"` or accept the new syntax. But we never heard any feedback about `babel-plugin-syntax-module-attributes` when renaming to `syntax-import-assertions` so I doubt it will break anyone.

**Comment by nicolo-ribaudo** [comment]
Sounds like a good plan to me 👍 
I'm not particularly worried about people using the old "module attributes" syntax, given that the syntax plugin has virtually 0 downloads (https://www.npmjs.com/package/@babel/plugin-syntax-module-attributes) and it has been deprecated for years.

**Comment by nicolo-ribaudo** [comment]
I managed to make it non-breaking also for users of the legacy syntax.
- By default, if the AST has `.assertions` we print `assert`
- By default, if the AST has `.attributes` we print the _legacy_ `with` syntax (without braces)

A generator option `importAttributesKeyword: "with" | "assert" | "with-legacy"` can control the output format. `@babel/plugin-syntax-import-attributes` will automatically set `importAttributesKeyword` to `"with"`, unless it has been explicitly set to something else.

If the AST has `.attributes` (thus if using `importAttributes`, `moduleAttributes`, or `importAssertions` with the `with` keyword) and `importAttributesKeyword` is not specified, `@babel/generator` will warn asking to explicitly set the option.

File: packages/babel-parser/src/parser/statement.ts [hunk 3]
```diff
@@ -3087,27 +3069,80 @@ export default abstract class StatementParser extends ExpressionParser {
       }
       node.value = this.parseStringLiteral(this.state.value);
       this.finishNode(node, "ImportAttribute");
-      attrs.push(node);
+      attrs.push(node as N.ImportAttribute);
```
Comment (JLHwung) [inline packages/babel-parser/src/parser/statement.ts:3072]: Nit, we can avoid the type cast:
```js
attrs.push(this.finishNode(node, "ImportAttribute"));
```

**Review by JLHwung — APPROVED** [review]
We should also handle the plugin assertions in `expression.js#finishCallExpression`

**Review by JLHwung — COMMENTED** [review]
Should also update how we calculate the `maxArgumentCount`.

**Comment by nicolo-ribaudo** [comment]
@JLHwung I don't know how I could forget to go through all the `hasPlugin("importAssertions")` calls to add the `hasPlugin("importAttributes")` variants where it makes sense 😅 I have done that now.

File: packages/babel-generator/src/generators/modules.ts [hunk 4]
```diff
@@ -66,15 +66,40 @@ export function ExportNamespaceSpecifier(
...
+      (this.format.importAttributesKeyword === "auto-legacy" && node.assertions)
+      ? "assert"
+      : "with",
+  );
   this.space();
+
+  if (
+    this.format.importAttributesKeyword === "with-legacy" ||
+    (this.format.importAttributesKeyword === "auto-legacy" && !node.assertions)
+  ) {
```
Comment (liuxingbaoyu) [inline packages/babel-generator/src/generators/modules.ts:95]: Can we have `importAttributesKeyword, assertions, attributes` as variables?

**Review by liuxingbaoyu — APPROVED** [review]

File: packages/babel-generator/src/printer.ts [hunk 5]
```diff
@@ -77,6 +77,15 @@ export type Format = {
    * @deprecated Removed in Babel 8
    */
   decoratorsBeforeExport?: boolean;
+  /**
+   * The import attributes syntax style:
+   * - "with"        : `import { a } from "b" with { type: "json" };`
+   * - "assert"      : `import { a } from "b" assert { type: "json" };`
+   * - "with-legacy" : `import { a } from "b" with type: "json";`
+   * - "auto-legacy" : "assert" if the AST node has an .assertions property,
+   *                   "with-legacy" if it has an .attributes property.
```
Comment (liuxingbaoyu) [inline packages/babel-generator/src/printer.ts:86]:
```suggestion
```

**Review by liuxingbaoyu — APPROVED** [review]
I like `IS_PUBLISH`, thanks!

**Comment by nicolo-ribaudo** [comment]
I'm merging this PR into a `feat-7.22.0/imports` branch, since there are different PRs related to changes in import statements.

**Pull request merged**